### ISSUE TITLE
Why is the large feature norm part of CTRL separated out and 5*torch.ones(img.shape) used as the label?

### ISSUE THREAD

**Issue by jsrdcht** [tardisblue9/DeDe#2]
Why is the large feature norm part of CTRL separated out and `5*torch.ones(img.shape)` used as the label?

Does this approach lead to information leakage? Does the decoder learn the specific pattern of CTRL poisons?

**Comment by tardisblue9** [comment]
CTRL is a special one due to its end-to-end setting. You can check _Auxiliary Explanation_ in _Evaluation_ for the reasoning.

For information leakage, we do not consider it as our threat model is to tackle backdoor attacks.

**Comment by jsrdcht** [comment]
In Auxiliary Explanation, the paper says
>filling the out-of-distribution regions with random noise image embeddings

It seems the code directly use the vanilla image embeddings and change the reconstruction target to `5*torch.ones(img.shape)`. 
```python
def ood_dataset_seperate(train_dataloader, args):
    if args.attack_type == 'ctrl':
        backdoored_encoder.eval()
        with torch.no_grad():
            feature_bank = []
            target_bank = []

            test_bar = tqdm(train_dataloader)
            for data, target, _ in test_bar:
                data, target = data.cuda(non_blocking=True), target.cuda(non_blocking=True)
                feature = backdoored_encoder(data)

                target_bank.append(target)
                feature_bank.append(feature)

            feature_bank = torch.cat(feature_bank, dim=0).contiguous()
            target_bank = torch.cat(target_bank, dim=0).contiguous()

        feature_bank = feature_bank.detach().cpu().numpy()
        target_bank = target_bank.detach().cpu().numpy()

        pca = PCA(n_components=2)
        result = pca.fit_transform(feature_bank)
        print(result.shape)
        label = target_bank[:result.shape[0]]
        df = pd.DataFrame(result)
        df['label'] = label
        sns.scatterplot(data=df, x=0, y=1, hue="label", marker="+")
        plt.savefig('./plot_result/ctrl_train_pca.jpg', dpi=260)
        plt.show()

        feature_norm = np.linalg.norm(feature_bank, ord=2, axis=1)
        df['norm'] = feature_norm
        threshold = df['norm'].quantile(1-args.poison_rate)  # .mean()+2
        print('threshold', threshold , "by quantile", 1-args.poison_rate)
        df_sub = df[df['norm'] > threshold]

        sns.scatterplot(data=df_sub, x=0, y=1, hue="label", marker="+")
        plt.savefig('./plot_result/ctrl_train_pca_rmood.jpg', dpi=260)
        plt.show()

        ood_dataset = torch.utils.data.Subset(shadow_data, df_sub.index) # here use the same image embeddings
        id_dataset = torch.utils.data.Subset(shadow_data, df[df['norm'] < threshold].index)
        print("after filter, we have {} id data and {} ood data".format(len(id_dataset), len(ood_dataset)))

        train_dataloader = DataLoader(id_dataset, batch_size=args.batch_size, shuffle=False, num_workers=16,
                                      pin_memory=True, drop_last=True)
        ood_train_dataloader = DataLoader(ood_dataset, batch_size=args.batch_size, shuffle=False, num_workers=16,
                                      pin_memory=True, drop_last=True)
        return train_dataloader, ood_train_dataloader
    else:
        return train_dataloader, None

def train_decoder_ood(backdoored_encoder, model, ood_data_loader, optimizer, args):
    backdoored_encoder.eval()
    model.train()
    if ood_data_loader == None:
        return 0

    total_loss, total_num, train_bar = 0.0, 0, tqdm(ood_data_loader)
    for batch in train_bar:
        img = batch[0]
        z_img = 5*torch.ones(img.shape).cuda() # torch.rand(img.shape).cuda() # here change the reconstruction target
        img = img.cuda()
        feature_raw = backdoored_encoder(img)
        predicted_img, mask = model(img, feature_raw)
        loss = torch.mean((predicted_img - z_img) ** 2 * mask) / args.mask_ratio
        optimizer.zero_grad()
        loss.backward()
        optimizer.step()

        total_loss += loss.item() * args.batch_size
        total_num += args.batch_size
        train_bar.set_description('Decoder Train Epoch: [{}/{}] ood Loss: {:.4f}'.format(epoch, args.epochs, total_loss / total_num))

    return total_loss / total_num
```
The paper and the code appear inconsistent. Have I misunderstood something?

**Comment by tardisblue9** [comment]
This line of code only serves as a special case for ctrl, so it is supposed to be commented out for any other attack run.

As for CTRL implementation, you can use `torch.rand(img.shape).cuda()` right behind in the comment to be consistent with the paper.

Also, as mentioned in the paper, a norm ball with radius of 0.99 quantiles should be in place to filter the outliers.

**Comment by jsrdcht** [comment]
OK, got you.
